### PR TITLE
Optional optic

### PR DESCRIPTION
**Pull request by nomisRev**
In OptionalLaws I made an non-reusable First tag. This could be replaced with a type tagging in Kategory. You don't get the same nice syntax as in Scala `Option<A> @@  First` but an type alias could always close the gap if required. Thoughts?

```
data class Tagged<A, T>(val value: A) : ReadOnlyProperty<Any?, A> {

    /**
     * Auto unwrapping by delegation
     */
    override fun getValue(thisRef: Any?, property: KProperty<*>): A = value

}

fun <A, T> A.tag(): Tagged<A, T> = Tagged(this)

sealed class UserToken

data class User(val name: String, val token: Tagged<String, UserToken>)

fun main(args: Array<String>) {


    val taggedString: Tagged<String, UserToken> = "someUserToken".tag()
    println("I was manually unwrapped: ${taggedString.value}")

    val string by taggedString
    println("I was auto unwrapped: $string")

    val user = User("John", taggedString)
    println("I can only be created with a UserToken tagged string: $user")

}
```

### CODE REVIEW

File: kategory-optics/src/main/kotlin/kategory/optics/Optional.kt [hunk 1]
```diff
@@ -0,0 +1,173 @@
+package kategory.optics
```
Comment (pakoito) [inline kategory-optics/src/main/kotlin/kategory/optics/Optional.kt]: package kategory

File: kategory-optics/src/test/kotlin/kategory/optics/IsoTest.kt [hunk 2]
```diff
@@ -3,7 +3,18 @@ package kategory.optics
 import io.kotlintest.KTestJUnitRunner
```
Comment (pakoito) [inline kategory-optics/src/test/kotlin/kategory/optics/IsoTest.kt]: package kategory :D

File: kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                }, a).value().option
+
+                getOption.exists { b ->
+                    modifyFId.exists {
+                        EQB.eqv(b, it)
+                    }
+                }
+            })
+
+    @PublishedApi internal data class FirstOption<A>(val option: Option<A>)
```
Comment (pakoito) [inline kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt]: What does @PublishedApi do? It's the first time I see it 😖

Comment (pakoito) [reply kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt]: Oh is this to allow inlines to use internals? Fuck yeah, I've been looking for something like this forever. We can hide some mechanisms now :D

File: kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            Law("Optional Law: modify identity = identity", { modifyIdentity(optional, aGen, EQA) }),
+            Law("Optional Law: compose modify", { composeModify(optional, aGen, funcGen, EQA) }),
+            Law("Optional Law: consistent set with modify", { consistentSetModify(optional, aGen, bGen, EQA) }),
+            Law("Optional Law: consistent modify with modify identity", { consistentModifyModifyId(optional, aGen, funcGen, EQA) }),
+            Law("Optional Law: consistent getOption with modify identity", { consistentGetOptionModifyId(optional, aGen, EQB) })
+    )
+
+    fun <A, B> getOptionSet(optional: Optional<A, B>, aGen: Gen<A>, EQA: Eq<A>): Unit =
+            forAll(aGen, { a ->
+                EQA.eqv(
```
Comment (pakoito) [inline kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt]: Is it worth using equalsUnderTheLaw here? We can get consistent changes in the future for all laws, and works like this behind the scenes.

**Comment by nomisRev** [comment]
@pakoito shall I make the changes here and merge or shall we wait on a decision on [polymorphic optics #277](https://github.com/kategory/kategory/pull/277)

File: kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt (hunk 4 above)
Comment (nomisRev) [reply kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt]: Yes, that should be `equalsUnderTheLaw`

**Comment by pakoito** [comment]
@nomisRev IMO we should merge this first, and continue the discussion about polymorphic lenses in the other ticket. I'm still wrapping my head around the concept, as Optics are somewhat new to me :D

**Pull request merged**